src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc

```diff
@@ -181,15 +155,12 @@ static void pf_cancel_picks_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol,
 static void start_picking_locked(grpc_exec_ctx *exec_ctx,
                                  pick_first_lb_policy *p) {
   p->started_picking = true;
-  if (p->subchannels != NULL) {
-    GPR_ASSERT(p->num_subchannels > 0);
-    p->checking_subchannel = 0;
-    p->checking_connectivity = GRPC_CHANNEL_IDLE;
-    GRPC_LB_POLICY_WEAK_REF(&p->base, "pick_first_connectivity");
-    grpc_subchannel_notify_on_state_change(
-        exec_ctx, p->subchannels[p->checking_subchannel],
-        p->base.interested_parties, &p->checking_connectivity,
-        &p->connectivity_changed);
+  if (p->subchannel_list != NULL && p->subchannel_list->num_subchannels > 0) {
+    p->subchannel_list->checking_subchannel = 0;
+    grpc_lb_subchannel_list_ref_for_connectivity_watch(
+        p->subchannel_list, "connectivity_watch+start_picking");
+    grpc_lb_subchannel_data_start_connectivity_watch(
```

Comment:
I can't find where `sd->pending_connectivity_state_unsafe` is set. There have been bugs due to having it unset, or having it set to IDLE (as in the previous version of this file). It needs to be set to the current connectivity state at the time of starting the watch.

Reply:
It doesn't need to be initialized.  The only place we ever look at it is in the connectivity callback, which won't be called until it's set.  Note that the RR code prior to this PR did not initialize it either.

Perhaps you're thinking of `prev_connectivity_state`?  That gets initialized in `grpc_lb_subchannel_list_create()`, the same way that it was initialized in RR before this PR.

Reply:
The docs of `grpc_subchannel_notify_on_state_change` say "call notify when the connectivity state of a channel changes from *state". That is, it's also part of the input, not just the output. If RR didn't previously do it, it was a bug. I remember similar bugs in the recent past: a subchannel went from INIT to CONNECTING without passing through IDLE (the default state, being int 0), resulting in the callback never being invoked.

Reply:
Then yeah, this is a pre-existing bug.  I've added a line in subchannel_list.cc to initialize it.